### PR TITLE
feat: create new job listing

### PR DESCRIPTION
**Pull request by Bola-coder**
### PR Description
**What does the PR do?**
This PR is about creating a new job endpoint.

How should this be manually tested?
Request Body
POST /api/jobs
`{
  "title": "Senior Developer",
  "description": "Responsible for developing and maintaining software applications.",
  "location": "Remote",
  "salary": "120000",
  "job_type": "Full-time"
}
`
**Successful Response**
On a successful creation of the job, the API should return a 201 Created status code.
The response body should contain the newly created job details.

`{
  "message": "Job created successfully",
  "job": {
    "title": "Senior Developer",
    "description": "Responsible for developing and maintaining software applications.",
    "location": "Remote",
    "salary": "120000",
    "job_type": "Full-time"
  }
}
`

**Authentication Error**
When the current user is not properly authenticated to create a job:

`{
  "status_code": 401,
  "error": "invalid authentication token",
  "message": "user not authorized for this action"
}
`

**Validation**
If any required field is missing or invalid, the response should return a 400 Bad Request status code with appropriate validation error messages:

`{
  "status_code": 400,
  "message": ["title must be a string", "salary must be a string"],
  "error": "Bad Request"
}
`

**Task**

- [x] Create a POST endpoint POST /api/v1/jobs/ to create a new job.
- [x] Implement server-side logic to handle successful and failed create requests.
- [x] Ensure the user has the proper permissions and authentication to create a job.
- [x] Write tests for creating job.
- [x] Update API documentation to reflect the new endpoint and its us

**Reference Issue**
[[FEAT]: ](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/159)

### CODE REVIEW

File: src/modules/job/job.service.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  constructor(
+    @InjectRepository(Job)
+    private jobRepository: Repository<Job>,
+    private organisationService: OrganisationsService // Inject OrganisationsService directly
+  ) {}
+
+  async createJob(job: CreateNewJobOption) {
+    const newJob = new Job();
+
+    // Find the organisation by its id or any other identifier
```
Comment (Cobidex) [inline src/modules/job/job.service.ts:20]: Remove this comment. It's no use here

File: src/modules/job/job.service.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import CreateNewJobOption from './options/CreateNewJobOption';
+import { Organisation } from '../organisations/entities/organisations.entity';
+import { OrganisationsService } from '../organisations/organisations.service';
+
+@Injectable()
+export class JobService {
+  constructor(
+    @InjectRepository(Job)
+    private jobRepository: Repository<Job>,
+    private organisationService: OrganisationsService // Inject OrganisationsService directly
```
Comment (Cobidex) [inline src/modules/job/job.service.ts:14]: Remove this comment

**Review by TheCodeGhinux — DISMISSED** [review]
Fair

File: src/modules/job/dto/create-job.dto.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { IsNotEmpty, isString } from 'class-validator';
```
Comment (Cobidex) [inline src/modules/job/dto/create-job.dto.ts:1]: Add documentation to your dto.for swagger..do this using the swagger decorators. For clarification reference the auth login and organisations module.

**Review by davidolanrewaju — DISMISSED** [review]

File: src/modules/job/tests/job.service.spec.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
+import { Test, TestingModule } from '@nestjs/testing';
+import { JobService } from '../job.service';
+import { Repository } from 'typeorm';
+import { Job } from '../entities/job.entity';
+import { getRepositoryToken } from '@nestjs/typeorm';
+import CreateNewJobOption from '../options/CreateNewJobOption';
+import { Organisation } from '../../organisations/entities/organisations.entity';
+import { OrganisationsService } from '../../organisations/organisations.service';
+
+// Mock repositories
```
Comment (Cobidex) [inline src/modules/job/tests/job.service.spec.ts:10]: It's obvious what's happening here. Comment is no use

File: src/modules/organisations/organisations.service.ts [hunk 5]
```diff
@@ -71,4 +71,8 @@ export class OrganisationsService {
       throw new InternalServerErrorException(`An internal server error occurred: ${error.message}`);
     }
   }
+
+  async findById(id: string): Promise<Organisation> {
```
Comment (Cobidex) [inline src/modules/organisations/organisations.service.ts:75]: Put a try/catch around your services or controllers where you are making database calls

File: src/modules/job/job.service.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  ) {}
+
+  async createJob(job: CreateNewJobOption) {
```
Comment (Cobidex) [inline src/modules/job/job.service.ts:17]: Try/catch block around areas of db calls. Ensure proper error login. Reference auth service for clarification

File: src/modules/organisations/tests/mocks/organisation.mock.ts [hunk 7]
```diff
@@ -41,6 +44,7 @@ export const createMockOrganisation = (): Organisation => {
     updated_at: new Date(),
     isDeleted: false,
     preferences: [],
+    jobs, // Add the jobs property
```
Comment (Cobidex) [inline src/modules/organisations/tests/mocks/organisation.mock.ts:47]: Dangling comment. Best removed

File: src/modules/organisations/tests/mocks/organisation.mock.ts [hunk 8]
```diff
@@ -25,6 +26,8 @@ export const createMockOrganisation = (): Organisation => {
     user_type: UserType.ADMIN,
   };
 
+  const jobs: Job[] = []; // Add an empty array or some mock jobs if needed
```
Comment (Cobidex) [inline src/modules/organisations/tests/mocks/organisation.mock.ts:29]: Dangling comment

**Comment by Bola-coder** [comment]
@Cobidex I have fixed all of those issues. Can you please help me tag a mentor to merge?

**Comment by Cobidex** [comment]
LGTM @buka4rill

File: src/modules/job/job.controller.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+@ApiBearerAuth()
+@ApiTags('Organisation')
+@Controller('jobs')
+export class JobController {
+  constructor(private readonly jobService: JobService) {}
+
+  @ApiOperation({ summary: 'Create Job Listing' })
+  @ApiResponse({
+    status: 201,
```
Comment (davidolanrewaju) [inline src/modules/job/job.controller.ts:14]: the convention to follow is status_code not status

File: src/modules/job/job.controller.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+import { ApiBearerAuth, ApiOperation, ApiResponse, ApiTags } from '@nestjs/swagger';
+
+@ApiBearerAuth()
+@ApiTags('Organisation')
+@Controller('jobs')
+export class JobController {
+  constructor(private readonly jobService: JobService) {}
+
+  @ApiOperation({ summary: 'Create Job Listing' })
+  @ApiResponse({
```
Comment (davidolanrewaju) [inline src/modules/job/job.controller.ts:13]: the convention to follow is status_code not status

Comment (Bola-coder) [reply src/modules/job/job.controller.ts:13]: I am not the one who created that. I only used it and tampering with that will cause a lot of conflicts

@davidolanrewaju

File: src/modules/job/job.service.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    try {
+      const newJob = new Job();
+
+      const organisation = await this.organisationService.findById(job.organisation);
+
+      if (!organisation) {
+        throw new Error('Organisation not found');
+      }
+
+      newJob.title = job.title;
```
Comment (Homoakin619) [inline src/modules/job/job.service.ts:27]: These multiple lines can be updated to about two using Object.assign

File: src/modules/organisations/organisations.controller.ts [hunk 12]
```diff
@@ -3,6 +3,8 @@ import { OrganisationsService } from './organisations.service';
 import { OrganisationRequestDto } from './dto/organisation.dto';
 import { ApiBearerAuth, ApiOperation, ApiResponse, ApiTags } from '@nestjs/swagger';
 import { UpdateOrganisationDto } from './dto/update-organisation.dto';
+import { skipAuth } from '../../../src/helpers/skipAuth';
```
Comment (Homoakin619) [inline src/modules/organisations/organisations.controller.ts:6]: this import was not used, kindly remove it

File: src/modules/organisations/organisations.service.ts [hunk 13]
```diff
@@ -71,4 +71,15 @@ export class OrganisationsService {
...
```
Comment (Homoakin619) [inline src/modules/organisations/organisations.service.ts:75]: Use more descriptive name like findOrganizationById

**Comment by Bola-coder** [comment]
I have fixed all @Homoakin619

File: src/modules/job/job.service.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    try {
+      const newJob = new Job();
+
+      const organisation = await this.organisationService.findOrganisationById(job.organisation);
+
+      if (!organisation) {
+        throw new Error('Organisation not found');
+      }
+
+      Object.assign(newJob, job);
```
Comment (Homoakin619) [inline src/modules/job/job.service.ts:27]: I hope you have tested this new implementation and confirmed it's not breaking anything

File: src/modules/organisations/organisations.service.ts [hunk 15]
```diff
@@ -71,4 +71,15 @@ export class OrganisationsService {
       throw new InternalServerErrorException(`An internal server error occurred: ${error.message}`);
     }
   }
+
+  async findOrganisationNyId(id: string): Promise<Organisation> {
```
Comment (Homoakin619) [inline src/modules/organisations/organisations.service.ts:75]: this should be findOrganisationById right? fix the typo error

**Comment by Bola-coder** [comment]
@Homoakin619 , The changes have been made., and yes, the new change works

**Review by Cobidex — APPROVED** [review]

**Comment by Bola-coder** [comment]
conflict fixed @Homoakin619 @TheCodeGhinux @Heba-WebDev @Cobidex

**Pull request closed without merging**